### PR TITLE
Merge persistence API for append history and conflict resolve

### PR DESCRIPTION
**Pull request by yiminc**
**What changed?**
Merge persistence API for append history and conflict resolve

**Why?**
Make it possible to optimize these operations in one transaction.

**How did you test it?**
Unit tests and integration tests.

**Potential risks**
No

**Is hotfix candidate?**
No

### CODE REVIEW

File: common/persistence/cassandra/cassandraPersistence.go [hunk 1]
```diff
@@ -1232,6 +1232,23 @@ func (d *cassandraPersistence) UpdateWorkflowExecution(
...
+		if err := d.AppendHistoryNodes(req); err != nil {
+			return err
+		}
+	}
+	for _, req := range request.CurrentWorkflowEventsNewEvents {
+		if err := d.AppendHistoryNodes(req); err != nil {
+			return err
+		}
+	}
+
```
Comment (wxing1292) [inline common/persistence/cassandra/cassandraPersistence.go:1251]: nit: can we do the current workflow events first, then reset, then new?

File: common/persistence/execution_manager.go [hunk 2]
```diff
@@ -120,31 +120,15 @@ func (m *executionManagerImpl) UpdateWorkflowExecution(
...
-	for _, workflowEvents := range request.NewWorkflowEvents {
-		newEvents, err := m.serializeWorkflowEvents(request.ShardID, workflowEvents)
-		if err != nil {
-			return nil, err
-		}
-		if newEvents != nil {
-			newWorkflowNewEvents = append(newWorkflowNewEvents, newEvents)
-			newWorkflowHistoryDiff += int64(len(newEvents.Node.Events.Data))
-		}
+	newWorkflowNewEvents, newWorkflowHistoryDiff, err := m.serializeWorkflowEventBatches(request.ShardID, request.NewWorkflowEvents)
```
Comment (wxing1292) [inline common/persistence/execution_manager.go:129]: no nil check? already done within?

File: common/persistence/execution_manager.go [hunk 3]
```diff
@@ -300,6 +310,26 @@ func (m *executionManagerImpl) CreateWorkflowExecution(
...
+	var workflowNewEvents []*InternalAppendHistoryNodesRequest
+	var historyDiff int64 = 0
+	for _, workflowEvents := range eventBatches {
+		newEvents, err := m.serializeWorkflowEvents(shardID, workflowEvents)
+		newEvents.ShardID = shardID
+		if err != nil {
+			return nil, 0, err
+		}
+		if newEvents != nil {
+			workflowNewEvents = append(workflowNewEvents, newEvents)
```
Comment (wxing1292) [inline common/persistence/execution_manager.go:326]: when will newEvents be nil?

Comment (wxing1292) [reply common/persistence/execution_manager.go:326]: can you make sure the newEvents is not nil?

from reading the code, newEvents is the serialized result of items within eventBatches, so if eventBatches does not contains nil element, this nil check is noop.

there should be no nil item in eventBatches in any condition (excluding bugs)
can you check why the nil check is necessary?

File: common/persistence/execution_manager.go [hunk 4]
```diff
@@ -300,6 +310,26 @@ func (m *executionManagerImpl) CreateWorkflowExecution(
 	return &CreateWorkflowExecutionResponse{HistorySize: newHistoryDiff}, nil
 }
 
+func (m *executionManagerImpl) serializeWorkflowEventBatches(
+	shardID int32,
+	eventBatches []*WorkflowEvents,
+) ([]*InternalAppendHistoryNodesRequest, int64, error) {
+	var workflowNewEvents []*InternalAppendHistoryNodesRequest
```
Comment (wxing1292) [inline common/persistence/execution_manager.go:317]:
```go
workflowNewEvents := make([]*InternalAppendHistoryNodesRequest, 0, len(eventBatches))
```

File: common/persistence/sql/execution.go [hunk 5]
```diff
@@ -481,6 +481,23 @@ func (m *sqlExecutionStore) updateWorkflowExecutionTx(
...
+	for _, req := range request.NewWorkflowEventsNewEvents {
+		if err := m.AppendHistoryNodes(req); err != nil {
+			return err
+		}
+	}
+	for _, req := range request.CurrentWorkflowEventsNewEvents {
+		if err := m.AppendHistoryNodes(req); err != nil {
+			return err
+		}
+	}
```
Comment (wxing1292) [inline common/persistence/sql/execution.go:499]: nil: can you put the current for loop at the very beginning?

File: common/persistence/cassandra/cassandraPersistence.go (hunk 1 above)
Comment (yiminc) [reply common/persistence/cassandra/cassandraPersistence.go:1251]: sure

Comment (yiminc) [reply common/persistence/cassandra/cassandraPersistence.go:1251]: what is the difference doing them in different order?

File: common/persistence/execution_manager.go (hunk 2 above)
Comment (yiminc) [reply common/persistence/execution_manager.go:129]: I could do this within the nil check, but then I have to declare those variable separately. The serializeWorkflowEventBatches handles nil as it uses for in range.

**Review by wxing1292 — APPROVED** [review]

File: common/persistence/cassandra/cassandraPersistence.go (hunk 1 above)
Comment (wxing1292) [reply common/persistence/cassandra/cassandraPersistence.go]: persistence layer logic usually first touch the current workflow first, see ref: https://github.com/temporalio/temporal/pull/1926/files#diff-f5746d819e8ff4338478e5ecf595530831afb175cc000f7258b82135a2c744ffR1265

**Pull request merged**